### PR TITLE
Added Shadow DOM awareness to selection

### PR DESCRIPTION
**Pull request by nazar-pc**
We need to use closest Shadow Root instead of Document for elements withing Shadow Root.
Otherwise we'll get no useful information about actual selection.

### CODE REVIEW

**Comment by nazar-pc** [comment]
Somebody alive here?

File: src/api/selection.js [hunk 1]
```diff
@@ -9,8 +9,21 @@ define(function () {
     // find the parent document or document fragment
     if( rootDoc.compareDocumentPosition(scribe.el) & Node.DOCUMENT_POSITION_DISCONNECTED ) {
       var currentElement = scribe.el.parentNode;
-      while(currentElement && nodeHelpers.isFragment(currentElement)) {
-        currentElement = currentElement.parentNode;
+      // Old browsers might not know about `:host` without polyfill, lets catch exception if it is the case
```
Comment (rrees) [inline src/api/selection.js]: I think this check should be in a different module

File: src/api/selection.js [hunk 2]
```diff
@@ -9,8 +9,21 @@ define(function () {
...
+      try {
+        var withinShadowRoot = currentElement.matches && currentElement.matches(':host *');
+      } catch (e) {}
+
+      // We need to use closest Shadow Root instead of Document for elements withing Shadow Root
+      // Otherwise we'll get no useful information about actual selection
```
Comment (rrees) [inline src/api/selection.js]: The conditional nesting is gnarly here. Have the while loop take a function and then use the `withinShadowRoot` flag to choose which terminating function to use.

**Review by rrees — COMMENTED** [review]
In some ways it feels like you should put Scribe into a ShadowDom mode in the options and then condition the behaviour off that.

Apart from comments though, looks sensible.

File: src/api/selection.js (hunk 1 above)
Comment (nazar-pc) [inline src/api/selection.js]: Do you mean AMD module here or what?

File: src/api/selection.js (hunk 2 above)
Comment (nazar-pc) [inline src/api/selection.js]: I agree, but don't quite understand how do you want it to be changed.
BTW, feel free to modify it if necessary (maintainers should be able to do that directly, otherwise send a PR against my fork).

File: src/api/selection.js (hunk 1 above)
Comment (rrees) [inline src/api/selection.js]: Yes, exactly. So this is the selection module so it feels like the wrong place for a Shadow DOM check, putting it in a Shadow DOM module and then importing it would make it easier to maintain.

File: src/api/selection.js (hunk 2 above)
Comment (rrees) [inline src/api/selection.js]: I've actually left the Guardian so I no longer have admin rights to the repo. This is kind of a drive-by reviewing.

I think the code should roughly be:

```
var/const/let rootCheck = isShadowDom(currentElement) ? shadowRootCheck : domRootCheck;

while(!rootCheck(currentElement)) {
  currentElement = currentElement.parentNode;
}
```

Make sense?
